substrate/frame/assets-holder/src/impls.rs

```diff
@@ -0,0 +1,265 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::*;
+
+use frame_support::traits::{
+	fungibles::{Dust, Inspect, InspectHold, MutateHold, Unbalanced, UnbalancedHold},
+	tokens::{
+		DepositConsequence, Fortitude, Precision, Preservation, Provenance, WithdrawConsequence,
+	},
+};
+use pallet_assets::BalanceOnHold;
+use sp_runtime::{
+	traits::{CheckedAdd, CheckedSub, Zero},
+	ArithmeticError,
+};
+
+// Implements [`BalanceOnHold`] from [`pallet-assets`], so it can understand whether there's some
+// balance on hold for an asset account, and is able to signal to this pallet when to clear the
+// state of an account.
+impl<T: Config<I>, I: 'static> BalanceOnHold<T::AssetId, T::AccountId, T::Balance>
+	for Pallet<T, I>
+{
+	fn balance_on_hold(asset: T::AssetId, who: &T::AccountId) -> Option<T::Balance> {
+		BalancesOnHold::<T, I>::get(asset, who)
+	}
+
+	fn died(asset: T::AssetId, who: &T::AccountId) {
+		BalancesOnHold::<T, I>::remove(asset.clone(), who);
+		Holds::<T, I>::remove(asset, who);
```

Comment:
we cannot do this without ensuring holds are empty or burning them, this will result incorrect total supply. please double check

Reply:
AFAIK, according to [this](https://github.com/paritytech/polkadot-sdk/pull/4530/files/aab6498e500041573120ba38fb6732cbfeefec9f#diff-1e9f9fedeee1bdf9064888c5dfb715d7efe303c1fc839663640013a11f82f302R203), an account `WouldDie` only if there's no `BalanceOnHold` nor `FrozenBalance` for that account.

This implies that, once `died` is called, it should be expected (not the same as it is expected) that these storage keys are already removed, and removing them here is just a mere failsafe.

That said… I can accept the proposal of burning the balance on hold once an account dies, if it somehow still exists.

Also, would gracefully fail in case the asset is not live (destroying, destroyed).

Reply:
> AFAIK, according to [this](https://github.com/paritytech/polkadot-sdk/pull/4530/files/aab6498e500041573120ba38fb6732cbfeefec9f#diff-1e9f9fedeee1bdf9064888c5dfb715d7efe303c1fc839663640013a11f82f302R203), an account WouldDie only if there's no BalanceOnHold nor FrozenBalance for that account.

then it should be part of the protocol, hence documented for `died` function in the trait. plus calling `defensive` function in this impl to have panic in test env and error log in prod should be enough. I think it is better for infallible function instead of `try_mutate`.

Reply:
Just changed it to use `defensive_assert!`.

Reply:
Why should we have a condition with `can_withdraw`? Asset or account should not be removed before holds cleaned up.

Reply:
Because assets can disappear, and by calling [`Assets::destroy_accounts`](https://github.com/pandres95/polkadot-sdk/blob/174b3d5fcab1cd15f825abbb98c29645f258c6ed/substrate/frame/assets/src/functions.rs#L819-L820), `died` would be called without the need of having this `WouldDie` consequence as a prerequisite to be called in first place.

So, a great way to know (because fetching the asset details is not possible outside the crate) whether the asset exists (and is live) is by using `can_withdraw`. It produces no output changes (internally it just calls `can_decreasea`) and retrieves a `WithdrawConsequence` that can be either `Success` or `UnknownAsset`. In the latter case, we know it's safe to remove the holds for this account because the asset is on its way of being destroyed (which is irreversible, anyways).

Reply:
I think it should not be possible to destroy account that has some freezes or holds. This would mean that any asset that has such account can not be destroyed. But if we do allow this any subsystem holding some freezes or holds might have invalid state. 
@joepetrowski what you think?

Reply:
Yeah that is tricky. I'm inclined to agree that an account with freezes/holds should not be destroyable. Presumably, if the asset issuer has retained control over the asset class, they have also retained control over uses of the asset (e.g. staking on a chain). Holds/freezes represent other systems using the asset, and those systems should need to be cleaned up before destroying the asset class that they depend on.

In the case that someone has privileged roles in the asset but not in the systems relying on the asset, well they shouldn't be able to just remove the asset that they all depend on.

Reply:
I agree with @joepetrowsky here. A solution I was exploring before is to either: a) make `died` return a Result, or b) ensure that an asset cannot start destroying if there are holds/freezes for such asset. This is something that can be inspected and returned to `pallet-assets` by extending a new method for the traits `BalanceOnHold` and `FrozenBalance`. Something like:

```rs
trait BalanceOnHold<AssetId> {
  fn can_destroy(id: AssetId) -> bool;
}
```

Reply:
@muharem @joepetrowski see this commit ([92769f1](https://github.com/paritytech/polkadot-sdk/pull/4530/commits/92769f1a825770364cf24cfa9a910903b7053897)) with a proposed implementation of that suggestion.

Reply:
I think we should just alter `dead_account` function and if `FrozenBalance::frozen_balance` or `BalanceOnHold:: balance_on_hold` not zero it should return `Keep`.

Reply:
Guess I wouldn't remove what's already there on `do_start_destroy`. Otherwise, there's the possibility that a [`defensive!`](https://github.com/pandres95/polkadot-sdk/blob/0e848ac34ec0162032a797ba58f28839f692119c/substrate/frame/assets/src/functions.rs#L819) call is hit.

Reply:
I took a look at `dead_account`. Not exactly sure if that's the best function to do that, as what it does is to evaluate the `DeadConsequence` based on an `ExistenceReason` (usually tied to a deposit).

At best, we could preserve what's already there (the `ContainsFreezes`/`ContainsHolds`), and prevent refunding the deposit from an asset account if these constraints still hold.

Reply:
yes, we should not refund if there is some freeze or hold, we should fail. it can be outside of `dead_account`, it is fine. but we should try to make it hard to forget this check before removing an account. you added those checks now but not in all places where we remove accounts (`decrease_balance`, `destroy_account`, `transfer_and_die`), also we even forget `died` in `transfer_and_die` function. this can be a proof for you it is easy now to make such mistake. additionally I think we should assert for `ContainsFreezes/ContainsHolds` after the pallet's checks like `NoDeposit/Unknown/IncorrectStatus/WouldBurn`.

Reply:
I see the point now. Maybe an idea is to make `dead_account` return a `Result` with either a `DeadConsequence` or an `Error`. That way, we don't forget, and there's a logical implication of either the account dies because of reason _x_, or fails to die because of error _y_.

Reply:
Applied on [25d9bb0](https://github.com/paritytech/polkadot-sdk/pull/4530/commits/25d9bb050df9347a71ad02e89b08f00ed4ba5400).

Reply:
I think we need tests for public functions trying/resulting to delete accounts with some holds and freezes. And this PR is ready for audit. 

Additionally it would be good to have the fix concerning freezes to the stable version that already released.

Reply:
Done on [fd3adef](https://github.com/paritytech/polkadot-sdk/pull/4530/commits/fd3adefd67d721c6bbb7f8c6650b140210526457).